cpp/src/io/json/json_column_csr.cu

```diff
@@ -0,0 +1,360 @@
+                 0);
+    thrust::for_each(rmm::exec_policy(stream),
+                     csr_unique_col_ids.begin(),
+                     csr_unique_col_ids.end(),
+                     [csr_column_categories = csr_column_categories.begin(),
+                      csr_parent_col_ids    = csr_parent_col_ids.begin(),
+                      csr_max_row_offsets   = csr_max_row_offsets.begin(),
+                      list_parents_children_max_row_offsets =
+                        list_parents_children_max_row_offsets.begin()] __device__(auto col_id) {
+                       auto csr_parent_col_id = csr_parent_col_ids[col_id];
+                       if (csr_parent_col_id != parent_node_sentinel and
+                           csr_column_categories[csr_parent_col_id] == node_t::NC_LIST) {
+                         cuda::atomic_ref<NodeIndexT, cuda::thread_scope_device> ref{
+                           *(list_parents_children_max_row_offsets + csr_parent_col_id)};
+                         ref.fetch_max(csr_max_row_offsets[col_id],
+                                       cuda::std::memory_order_relaxed);
+                       }
+                     });
+    thrust::gather_if(
+      rmm::exec_policy(stream),
+      csr_parent_col_ids.begin(),
+      csr_parent_col_ids.end(),
+      csr_parent_col_ids.begin(),
+      list_parents_children_max_row_offsets.begin(),
+      csr_max_row_offsets.begin(),
+      [csr_column_categories = csr_column_categories.begin()] __device__(size_type parent_col_id) {
+        return parent_col_id != parent_node_sentinel and
+               csr_column_categories[parent_col_id] == node_t::NC_LIST;
+      });
+  }
+
+  // copy lists' max_row_offsets to children.
+  // all structs should have same size.
+  thrust::transform_if(
+    rmm::exec_policy(stream),
+    csr_unique_col_ids.begin(),
+    csr_unique_col_ids.end(),
+    csr_max_row_offsets.begin(),
+    [csr_column_categories = csr_column_categories.begin(),
+     is_non_list_parent,
+     csr_parent_col_ids  = csr_parent_col_ids.begin(),
+     csr_max_row_offsets = csr_max_row_offsets.begin()] __device__(size_type col_id) {
+      auto parent_col_id = csr_parent_col_ids[col_id];
+      // condition is true if parent is not a list, or sentinel/root
+      while (is_non_list_parent(parent_col_id)) {
+        col_id        = parent_col_id;
+        parent_col_id = csr_parent_col_ids[parent_col_id];
+      }
+      return csr_max_row_offsets[col_id];
+    },
+    [csr_column_categories = csr_column_categories.begin(),
+     is_non_list_parent,
+     parent_col_ids = csr_parent_col_ids.begin()] __device__(size_type col_id) {
+      auto parent_col_id = parent_col_ids[col_id];
+      // condition is true if parent is not a list, or sentinel/root
+      return is_non_list_parent(parent_col_id);
+    });
+
+  // For Struct and List (to avoid copying entire strings when mixed type as string is enabled)
+  thrust::transform_if(
```

Comment:
Overall, I'm terrible by the amount of thrust calls in this function :fearful: . It is not our issue, just the algorithm issue. I'm not sure how fast the entire function can run.